### PR TITLE
feat(meta): optimize table stats report performance

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close https://github.com/risingwavelabs/risingwave/issues/15153

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/hummock/metrics_utils.rs [hunk 1]
```diff
@@ -37,6 +39,72 @@ use crate::hummock::checkpoint::HummockVersionCheckpoint;
...
+    m
+}
+pub fn trigger_table_stat(
+    metrics: &MetaMetrics,
+    local_metrics: &mut HashMap<u32, LocalTableMetrics>,
+    version_stats: &HummockVersionStats,
+    table_stats_change: &PbTableStatsMap,
+) {
+    for (table_id, stats) in table_stats_change {
+        if stats.total_key_size == 0 && stats.total_value_size == 0 && stats.total_key_count == 0 {
```
Comment (Li0k) [inline src/meta/src/hummock/metrics_utils.rs:80]: Shall we need to check `0` after `purge_prost_table_stats` ?

File: src/meta/src/hummock/manager/mod.rs [hunk 2]
```diff
@@ -1430,13 +1432,18 @@ impl HummockManager {
...
                     VarTransaction::new(&mut versioning.version_stats,)
                 );
-                if let Some(table_stats_change) = &table_stats_change {
-                    add_prost_table_stats_map(&mut version_stats.table_stats, table_stats_change);
-                }
 
                 // apply version delta before we persist this change. If it causes panic we can
                 // recover to a correct state after restarting meta-node.
                 current_version.apply_version_delta(&version_delta);
+                if purge_prost_table_stats(&mut version_stats.table_stats, &current_version) {
```
Comment (Li0k) [inline src/meta/src/hummock/manager/mod.rs:1439]: IMO, This metric is just an approximation, we don't need to update it every time we report the compact task, we can reduce the overhead by reporting it less frequently, what do you think?

Comment (Little-Wallace) [reply src/meta/src/hummock/manager/mod.rs:1446]: This PR only report the table which has update

**Review by hzxa21 — COMMENTED** [review]
There are three differnet logics to report the table stats metrics in differnet places:
1. init: `get_local_table_stats`
2. report_compact_task: `trigger_local_table_stat`
3. commit_epoch: `trigger_table_stat`

This looks ad-hoc and not clean. Some ideas to improve:
- Maintain local table stats (with table_id -> Int instead of IntGauge) under versioning
- Update the local table stats under version write lock (which is already acquired) in report_compact_task and commit_epoch. We can do it incrementally and no reset is needed.
- Report local table stats to prometheus in HummockTimerEvent::Report

**Comment by Little-Wallace** [comment]
> There are three differnet logics to report the table stats metrics in differnet places:
> 
> 1. init: `get_local_table_stats`
> 2. report_compact_task: `trigger_local_table_stat`
> 3. commit_epoch: `trigger_table_stat`
> 
> This looks ad-hoc and not clean. Some ideas to improve:
> 
> * Maintain local table stats (with table_id -> Int instead of IntGauge) under versioning
> * Update the local table stats under version write lock (which is already acquired) in report_compact_task and commit_epoch. We can do it incrementally and no reset is needed.
> * Report local table stats to prometheus in HummockTimerEvent::Report

I have refactor code to maintain local table stats (with table_id -> Int instead of IntGauge) under versioning.

**Comment by hzxa21** [comment]
> > There are three differnet logics to report the table stats metrics in differnet places:
> > 
> > 1. init: `get_local_table_stats`
> > 2. report_compact_task: `trigger_local_table_stat`
> > 3. commit_epoch: `trigger_table_stat`
> > 
> > This looks ad-hoc and not clean. Some ideas to improve:
> > 
> > * Maintain local table stats (with table_id -> Int instead of IntGauge) under versioning
> > * Update the local table stats under version write lock (which is already acquired) in report_compact_task and commit_epoch. We can do it incrementally and no reset is needed.
> > * Report local table stats to prometheus in HummockTimerEvent::Report
> 
> I have refactor code to maintain local table stats (with table_id -> Int instead of IntGauge) under versioning.

Did you forget to push?

File: src/meta/src/hummock/manager/tests.rs [hunk 3]
```diff
@@ -265,7 +265,7 @@ async fn test_hummock_compaction_task() {
     // Finish the task and succeed.
 
     assert!(hummock_manager
-        .report_compact_task(compact_task.task_id, TaskStatus::Success, vec![], None)
+        .report_compact_task(compact_task.task_id, TaskStatus::Success, vec![], None,)
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/tests.rs:268]: remove `,` at the end

File: src/meta/src/hummock/manager/tests.rs [hunk 4]
```diff
@@ -1972,7 +1972,7 @@ async fn test_compaction_task_expiration_due_to_split_group() {
     let version_1 = hummock_manager.get_current_version().await;
     // compaction_task.task_status = TaskStatus::Success.into();
     assert!(!hummock_manager
-        .report_compact_task(compaction_task.task_id, TaskStatus::Success, vec![], None)
+        .report_compact_task(compaction_task.task_id, TaskStatus::Success, vec![], None,)
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/tests.rs:1975]: remove `,` at the end

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

**Pull request merged**